File: leetcode/Combination_Sum.java
import java.util.*;

public class Solution {
    void search(int[] cand, int target, int cur, int sum, ArrayList<Integer> in, ArrayList<ArrayList<Integer>> ret) {
        if (sum == target) {
            ret.add(new ArrayList<Integer>(in));
            return;
        }
        for (int i = cur; i < cand.length; ++i)
            if (sum + cand[i] <= target) {
                in.add(cand[i]);
                search(cand, target, i, sum + cand[i], in, ret);
                in.remove(in.size() - 1);
            }
    }
    public ArrayList<ArrayList<Integer>> combinationSum(int[] candidates, int target) {
        ArrayList<ArrayList<Integer>> ret = new ArrayList<ArrayList<Integer>>();
        Arrays.sort(candidates);
        ArrayList<Integer> in = new ArrayList<Integer>();
        search(candidates, target, 0, 0, in, ret);
        return ret;
    }
}

File: leetcode/Combination_Sum.cpp
class Solution {
    vector<vector<int> > combinations;
    void generate(int k, vector<int> &candidates, int sum, 
                  vector<int> &combination, int target) {
        if (k == candidates.size()) {
            if (sum == target)
                combinations.push_back(combination);
            return;
        }
        for (int i = 0; sum + candidates[k] * i <= target; ++i) {
            for (int j = 1; j <= i; ++j)
                combination.push_back(candidates[k]);
            generate(k + 1, candidates, sum + candidates[k] * i, 
                     combination, target);
            for (int j = 1; j <= i; ++j)
                combination.pop_back();
        }
    }
public:
    vector<vector<int> > combinationSum(vector<int> &candidates, int target) {
        combinations.clear();
        sort(candidates.begin(), candidates.end());
        vector<int> combination;
        generate(0, candidates, 0, combination, target);
        return combinations;
    }
};
